### ISSUE TITLE
Auto-BCC Option

### ISSUE THREAD

**Issue by Braintelligence** [mailcow/mailcow-dockerized#234]
Hi there!
In Webmin it is possible to choose in a binary way if a mailbox should receive mail to itself and/or forward to another address (alias) and/or send an automatic reply:
![grafik](https://cloud.githubusercontent.com/assets/6761196/25583626/ef01958e-2e92-11e7-826d-ce70de7bd53e.png)

What is the best way in mailcow to have a mailbox that receives its own mail but also forwards it to another address?

**Comment by mkuron** [comment]
Set up a Sieve rule: log in to SOGo with that account, Preferences, Mail, Filters and create a rule that forwards all messages to the desired email address. You can also have it send an automatic reply.

**Comment by Braintelligence** [comment]
So just to clarify:
If I want a mail address to have its own mailbox but to also forward it to another mail address I will have to log in to SOGo for each of those and set this up?

Then another question:
What happens in this case when the mailbox quota is reached? Is the mail still being forwarded then but just not delivered to the mailbox?

**Comment by mkuron** [comment]
Yes, you need to create a mailbox for each, log in and set up the rule.

Sieve rules are applied before the message is stored, so the forward should happen even if the mailbox is full. Please test this though -- just set the quota to 1 MB and send a message with a larger attachment to that mailbox.

**Comment by Braintelligence** [comment]
I will, after I moved all my mailboxes 😸.

**Comment by Braintelligence** [comment]
@mkuron Thank you very much for your help! Another question came up: In SOGo I can choose to forward and to keep a copy.
If I untick keeping a copy will all mails only be forwarded without keeping them in the mailbox?

**Comment by mkuron** [comment]
Yes, that would probably do it. But if you don't want to keep the messages, you could just create an alias instead of a mailbox.

The "Forward" setting in SOGo translates into a sieve rule like `redirect "test@example.com;"`; ticking the checkbox adds `keep;` after that.

**Comment by Braintelligence** [comment]
I want some addresses to be real users/mailboxes, so external apps can use them via SMTP.
For the receiving end I only want them to forward, though.

Is this use-case alright? 😸

**Comment by mkuron** [comment]
Ok, makes sense.

**Comment by Braintelligence** [comment]
It would be awesome, if mailcow supported this very use-case when you create mailboxes.
Just a tiny textfield saying forward to address and a checkbox if a copy should be kept. 😸

**Comment by Braintelligence** [comment]
@mkuron Can you point me towards how to set up a mailbox that copies all outgoing mail into its inbox? In webmin you could set it via an option named "BCC Mapping".

**Comment by chaosbunker** [comment]
I also think it would be cool if it was possible to set up mail forwarding as described by @Braintelligence in mailcow UI without having to touch SOGo

**Comment by mkuron** [comment]
@Braintelligence, that "BCC Mapping" would have to be done by modifying Postfix configuration ([`sender_bcc_maps`](http://www.postfix.org/postconf.5.html#sender_bcc_maps)). Outgoing messages never go through Dovecot/Sieve, so a rule in SOGo won't help.

**Comment by Braintelligence** [comment]
Is there an option for this in the mailcow ui, though? 😸 
If not it's not that big of a deal.
Can you point me towards how to map a whole domain to a single mailbox for outgoing bcc? 😺

**Comment by mkuron** [comment]
There is no GUI option for that, but adding the following line to data/conf/postfix/main.cf would probably work for your use case:
```
sender_bcc_maps = inline:{ @mydomain.com=archive@mydomain.com }
```

**Comment by Braintelligence** [comment]
Awesome, @mkuron you're the best!
I think all my main use-cases are dealt with now, so I'm closing this, or should we keep this open as a feature request for inline creation of mailboxes with forwarding?

**Comment by mkuron** [comment]
I'm leaving this open as a feature request. It's relatively trivial to implement Auto-BCC into the web UI, so if somebody needs it they should to it and submit a PR.

**Comment by develth** [comment]
Hi!

I´m using a mail archive tool (piler) to make sure my mails get archived. 
For that i have to set for some domains bcc maps and also a `x-add-envelope-to`.

Currently i solve it on my own mailserver setup with 

    recipient_bcc_maps = hash:/etc/postfix/bcc_maps
    sender_bcc_maps =  hash:/etc/postfix/bcc_maps

In the maps are entries like that:

      @domain1.com                archiv@myarchive.com
      @domain-two.com          archiv@myarchive.com

If there are other wo need a feature like this and my current setup would also fit this enhancement, i will have a look and try to make a PR. 

Cheers, 
Thomas

**Comment by re-host** [comment]
@develth I cant program that, but really would appreciate if you could make a PR :) would also be awesome if this feature would come into the Master

**Comment by Braintelligence** [comment]
@develth Are you using piler via docker? If you implemented it within the docker-compose.yml of mailcow, I would be very happy to hear how you did it. See also https://github.com/mailcow/mailcow-dockerized/issues/457

**Comment by develth** [comment]
@Braintelligence no - it runs on a different secured server.

@re-host for sure, when i got time for it ;)

@andryyy how did you integrated piler? via a custom postfix config?

**Comment by Braintelligence** [comment]
@develth I see, do you use it in a dockerized fashion on that other server or "natively"?

**Comment by develth** [comment]
@Braintelligence currently native.

**Comment by Braintelligence** [comment]
@develth Is any form of special hardware involved, or is it just meant for "normal" backups? (Last question, sorry for nagging 😸)

**Comment by develth** [comment]
@Braintelligence special hardware and special tools (e.g. bacula) - now back to topic ;)

**Comment by andryyy** [comment]
@develth In hosted mailcow? Pretty much the same. Hide the bcc fields, send copy to Piler. Piler sits on a different server with an audit log.

**Comment by alexanderrieder** [comment]
any news on this? :)

**Comment by mkuron** [comment]
It was implemented on the dev branch a few weeks ago: ade4b9e7ae7e28535ed261535526bc4e98ff48b4. Should be coming soon.
